torch/nested/_internal/ops.py

```diff
@@ -0,0 +1,156 @@
+import torch
+import functools
+from .nested_tensor import NestedTensor, is_jagged
+from typing import *  # noqa: F403
+
+__all__ = []
+
+JAGGED_OPS_TABLE = {}
+
+def check_schema(schema_str: str, func, *args, **kwargs) -> None:
+    named_arg_types = schema_str.split(", ")
+    num_optional_args = sum([x.endswith("?") for x in named_arg_types])
+    min_args = len(named_arg_types) - num_optional_args
+
+    if not (len(args) >= min_args and len(args) <= len(named_arg_types)):
+        raise ValueError(
+            f"NestedTensor {func.__name__}({schema_str}): expected at least {min_args} "
+            f"arguments and at most {len(named_arg_types)} arguments, but got: "
+            f"{len(args)} arguments")
+
+    arg_type_check_fns = {
+        "t": lambda x: isinstance(x, torch.Tensor) and not isinstance(x, NestedTensor),
+        "jt": is_jagged,
+        "any": lambda x: True,
+    }
+    for i, named_arg_type in enumerate(named_arg_types):
+        name, arg_type = named_arg_type.split(": ")
+        is_optional = arg_type.endswith("?")
+        normalized_arg_type = arg_type[:-1] if is_optional else arg_type
+        if not normalized_arg_type in arg_type_check_fns.keys():
+            raise AssertionError(f"Unknown arg type: {normalized_arg_type}")
+
+        if i >= len(args):
+            if not is_optional:
+                raise ValueError(f"NestedTensor {func.__name__}({schema_str}) "
+                                 f"missing required argument: {name}")
+            continue
+
+        if not arg_type_check_fns[normalized_arg_type](args[i]):
+            raise ValueError(
+                f"NestedTensor {func.__name__}({schema_str}): {name} should be of "
+                f"type {arg_type}, but got: {type(args[i])}")
+
+def check_ragged_dim_same(
+        func,
+        a: NestedTensor,
+        a_name: str,
+        b: NestedTensor,
+        b_name: str) -> None:
+    # Calling into .shape here
+    assert len(a._size) == 3, "NestedTensor must be [B, *, D]"
```

Comment:
note: we'll need to support `[B, *, D_0, ..., D_N]` for SAM. we have at least N=1